### PR TITLE
Sqlalchemy adapter

### PR DESCRIPTION
**Pull request by davizucon**
This is the very first release of Sqlalchemy adapter, that have SQLite as default implementation, remains in **alpha**.

- Notes: 
   - Tables may change (and will), so, save your training data. There no data migration (yet, not for alpha).
   - Performance test not done yet.
   - Tests using others databases not finished.
   - Documentation on the way, any question, go to  #241  or chat on gitter. (faster with mention)

Thanks,
Davi

### CODE REVIEW

**Comment by davizucon** [comment]
Finally !

File: chatterbot/storage/sqlalchemy_storage.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        #                        Column('in_response_to', Integer, ForeignKey('response.id')),
+        #                        Column('extra_data', Text)
+        #                        )
+        #
+        # # mapper(Response, self.response,
+        # #        # non_primary=True,
+        # #        properties={
+        # #            'statement': relationship(Statement, backref='response')
+        # #        }, )
+        # mapper(Statement, self.statement)
```
Comment (vkosuri) [inline chatterbot/storage/sqlalchemy_storage.py]: I think you could remove these lines?

File: chatterbot/storage/sqlalchemy_storage.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            if isinstance(statement, ResponseTable):
+                if statement and statement.statement_table:
+                    results.append(statement.statement_table.get_statement())
+            else:
+                if statement:
+                    results.append(statement.get_statement())
+
+        return results
+
+    # def filter(self, **kwargs):
```
Comment (vkosuri) [inline chatterbot/storage/sqlalchemy_storage.py]: I think you could remove these lines?

File: requirements.txt [hunk 3]
```diff
@@ -3,3 +3,5 @@ jsondatabase>=0.1.7,<1.0.0
 nltk>=3.2.0,<4.0.0
 pymongo>=3.3.0,<4.0.0
 python-twitter>=3.0.0,<4.0.0
+textblob>=0.11.0,<0.12.0
```
Comment (vkosuri) [inline requirements.txt]: Did you imported this package? I haven't seen

File: chatterbot/storage/__init__.py [hunk 4]
```diff
@@ -2,3 +2,9 @@
 from .django_storage import DjangoStorageAdapter
 from .jsonfile import JsonFileStorageAdapter
 from .mongodb import MongoDatabaseAdapter
+
+# FIXME Better way manage import
+try:
```
Comment (vkosuri) [inline chatterbot/storage/__init__.py]: I think these files are static and reside in same directory as long as they are. It's always a safe import.

File: chatterbot/storage/sqlalchemy_storage.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+    # Test relationship:
+    statement_table = relationship("StatementTable", back_populates="in_response_to", cascade="all", uselist=False)
+    text_search = Column(String, primary_key=True, default=get_reponse_serialized)
+
+    def get_response(self):
+        occ = {"occurrence": self.occurrence}
+        return Response(text=self.text, **occ)
+
+
+# # relational table  TO REMOVE
```
Comment (vkosuri) [inline chatterbot/storage/sqlalchemy_storage.py]: I think you could remove these lines

**Comment by davizucon** [comment]
Hi @vkosuri , Indeed.

**Comment by addelll** [comment]
there is any way to see exemple using sqlalchemy ?

**Comment by gunthercox** [comment]
Hi @addelll, you can use this storage adapter the same way that the others are used.

For example,

```python
chatbot = ChatBot(
    "My ChatterBot",
    storage_adapter="chatterbot.storage.SQLAlchemyDatabaseAdapter"
)
```

Is this what you are looking for as an example?

**Comment by addelll** [comment]
AttributeError: module 'chatterbot.storage' has no attribute 'SQLAlchemyDatabaseAdapter' ???

**Comment by gunthercox** [comment]
The reason you are seeing that message is likely because the current version of ChatterBot (`0.6.0`) doesn't contain the changes that were added in this pull request.

**Comment by addelll** [comment]
so any suggestions or i must copy directly from github ??

**Comment by gunthercox** [comment]
One possible way is to install directly from GitHub.

```
pip install git+git://github.com/gunthercox/ChatterBot.git@master
```

**Comment by addelll** [comment]
ok some errors 
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: StatementTable

**Comment by gunthercox** [comment]
@addelll I believe you will need to set `create=True` when using this adapter for the first time.

**Comment by davizucon** [comment]
@gunthercox @addelll , Sorry, I'll update documentation, asap.

**Comment by addelll** [comment]
@gunthercox @davizucon thanks for the greate work , i need one full exemple using Sqlalchemy adapter because i have some errors after second release in heroku like 
**_Arguments: ('columns text, text_search are not unique',)_**
I tried to fix that but ....

**Pull request merged**